poppy/wavefront.py

```diff
@@ -0,0 +1,359 @@
+    
+    The extension includes the redefinition of the wavefront array to
+    represent an electric field with the appropriate units V.m^-1.
+    Some utility functions have been defined to be able to compute the beam
+    radius and beam quality factor. Additionally, a wavelength scaling has 
+    been introduced to simulate a beam quality factor.
+    
+    Parameters
+    ----------
+    M2 : float
+        Beam quality factor.
+    n0 : float
+        Ambient refractive index.
+    """
+    
+    def __init__(self,
+                 beam_radius,
+                 units=u.m,
+                 rayleigh_factor=2.0,
+                 oversample=2,
+                 wavelength=1064.0e-9,
+                 npix=1024,
+                 M2=1.0,
+                 n0=1.00027398, # refractive index of air @ 15°C, lambda=1064nm
+                 **kwargs):
+        super(PhysicalFresnelWavefront, self).__init__(
+                beam_radius=beam_radius,
+                units=units,
+                rayleigh_factor=rayleigh_factor,
+                oversample=oversample,
+                wavelength=wavelength*M2/n0, # scale the wavelength and apply refractive index
+                npix=npix,
+                **kwargs
+                )
+        self.lam = wavelength/n0 # its unscaled value is needed to compute M2
+        self.n0 = n0
+        self.npix = npix*oversample
+        
+        return
+    
+    @property
+    def x(self):
+        """Spatial grid (m)."""
+        y, x = self.coordinates()
+        
+        return x[0, :]
+    
+    @property
+    def dx(self):
+        """Spatial grid sampling (m)."""
+        
+        return self.pixelscale.to(u.m/u.pixel).value
+    
+    @property
+    def xy(self):
+        """Spatial grid (m) as 2 dimensional array."""
+        y, x = self.coordinates()
+        
+        return np.array([x.flatten(), y.flatten()]).T
+        
+    @property
+    def q(self):
+        """Momentum grid (m^-1)."""
+        
+        return np.fft.fftfreq(self.npix, d=self.dx)*2.0*np.pi
+    
+    @property
+    def dq(self):
+        """Momentum grid sampling (m^-1)."""
+        
+        return 2.0*np.pi/self.npix/self.dx
+    
+    @property
+    def intensity(self):
+        """Intensity distribution (W.m^-2)."""
+        
+        return const.c * self.n0 * const.epsilon_0 * np.abs(self.amplitude)**2 / 2.0
+    
+    @property
+    def power(self):
+        """Power of the wavefront (W)."""
+        
+        intensity = self.intensity
+        dx = self.dx
+        
+        return dx**2*np.sum(intensity)
+    
+    def scale(self, P):
+        """
+        Scales the wavefront to a desired power.
+        
+        Parameters
+        ----------
+        P : float
+            The desired power of the wavefront.
+        """
+        
+        P0 = self.power
+        self.wavefront *= np.sqrt(P/P0)
+        
+        return
+    
+    def propagate(self, z, attenuation_coeff=0.0, **kwargs):
```

Comment:
Let's try to keep the API consistent between this and the parent class. Instead of adding a new function named `propagate`, this should just be named `propagate_fresnel`. Then it would call the parent class function of the same name via `super()`:
```
        P = self.power
        super(PhysicalFresnelWavefront,self).propagate_fresnel(z, **kwargs)
        self.scale_power(P*np.exp(-attenuation_coeff*z.to(u.m).value))
```

Doing it that way would ensure that all propagations of the `PhysicalFresnelWavefront` would use the power-conserving propagation, instead of letting someone accidentally call the non-power-conserving one.